### PR TITLE
WIP: Add storage recommendation

### PR DESCRIPTION
**Pull request by katrybacka**

### CODE REVIEW

**Comment by katrybacka** [comment]
recheck no bug

**Comment by katrybacka** [comment]
TODO (before merge): Add VC_STORAGE_POD environment variable to Jenkins jobs

File: ansible/files/provision_vm.py [hunk 1]
```diff
@@ -14,6 +14,11 @@ def get_args():
                         action='store',
                         help='Cluster to use (if not provided, script will choose the first one available)')
 
+    parser.add_argument('--storage-pod',
```
Comment (sodar) [inline ansible/files/provision_vm.py]: I would prefer if this parameter was named `datastore-cluster`.

File: ansible/files/vmware_api.py [hunk 2]
```diff
@@ -239,6 +227,52 @@ def get_vm_clone_spec(config_spec, customization_spec, relocate_spec):
...
+    storage_spec.cloneName = name
+    storage_spec.folder = folder
+    storage_spec.podSelectionSpec = pod_selection_spec
+    storage_spec.vm = vm
+    storage_spec.cloneSpec = clone_spec
+    storage_spec.type = operation_type
+    return storage_spec
+
+def get_vm_pod_selection_spec(api, storage_pod_name):
+    storage_pod = _get_storage_pod(api, storage_pod_name)
```
Comment (sodar) [inline ansible/files/vmware_api.py]: We should check if `_get_storage_pod` returns `None` and raise an exception if it does.

File: ansible/files/vmware_api.py [hunk 3]
```diff
@@ -239,6 +227,52 @@ def get_vm_clone_spec(config_spec, customization_spec, relocate_spec):
...
+    pod_selection_spec.storagePod = storage_pod
+    return pod_selection_spec
+
+
+def _get_storage_pod(api, storage_pod_name):
+    storage_pod = api.get_vc_object(vim.StoragePod, storage_pod_name)
+    return storage_pod
+
+
+def get_apply_storage_recommendation_task(api, storage_spec):
```
Comment (sodar) [inline ansible/files/vmware_api.py]: Maybe `clone_template_to_datastore_cluster` would be a better name for this function.

File: ansible/files/vmware_api.py [hunk 4]
```diff
@@ -239,6 +227,52 @@ def get_vm_clone_spec(config_spec, customization_spec, relocate_spec):
...
+    storage_manager = _get_storage_resource_manager(api)
+    recommendation_key = _get_recommended_datastore_key(storage_manager, storage_spec)
+    task = storage_manager.ApplyStorageDrsRecommendation_Task(key = recommendation_key)
+    return task
+
+
+def _get_storage_resource_manager(api):
+    storage_manager = api.content.storageResourceManager
```
Comment (sodar) [inline ansible/files/vmware_api.py]: We can cache `storage_manager` object in `VmwareApi` class, just like `content` or `datacenter` objects. We can do it in `_setup_common_objects` method.

File: ansible/files/vmware_api.py [hunk 5]
```diff
@@ -239,6 +227,52 @@ def get_vm_clone_spec(config_spec, customization_spec, relocate_spec):
...
+
+def _get_recommended_datastore_key(storage_manager, storage_spec):
+    recommended_datastore = _get_recommended_datastore(storage_manager, storage_spec)
+    recommendation_key = recommended_datastore.key
+    return recommendation_key
+
+
+def _get_recommended_datastore(storage_manager, storage_spec):
+    recommended_datastores = storage_manager.RecommendDatastores(storageSpec = storage_spec)
+    recommended_datastore = recommended_datastores.recommendations[0]
```
Comment (sodar) [inline ansible/files/vmware_api.py]: Please check if `recommendations` is list has any elements and throw as exception if it does not.

File: ansible/files/vmware_api.py [hunk 6]
```diff
@@ -239,6 +227,52 @@ def get_vm_clone_spec(config_spec, customization_spec, relocate_spec):
...
+    task = storage_manager.ApplyStorageDrsRecommendation_Task(key = recommendation_key)
+    return task
+
+
+def _get_storage_resource_manager(api):
+    storage_manager = api.content.storageResourceManager
+    return storage_manager
+
+
+def _get_recommended_datastore_key(storage_manager, storage_spec):
```
Comment (sodar) [inline ansible/files/vmware_api.py]: I think we can just return `recommended_datastores.recommendations[0].key` in `_get_recommended_datastore`. There is no need for a separate function for extracting key.
Them we could rename `_get_recommended_datastore` to `_get_recommendation_key`.

**Pull request merged**